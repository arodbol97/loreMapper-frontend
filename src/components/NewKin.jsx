import React, { useState, useEffect } from 'react';
import axios from 'axios';

const NewKin = ({ worldId, displayMarkers, setShowNewForm , getKins }) => {
    const divStyles = {
        display: 'flex',
        flexWrap: 'wrap',
        width: '100%',
        alignItems: 'flex-start'
    };

    const formStyles = {
        width: '100%',
    };

    const labelStyles = {
        color: 'white',
        width: '20%',
        textAlign: 'right',
        paddingRight: '10px',
        marginBottom: '5px'
    };

    const inputStyles = {
        width: 'calc(100% - 0px)',
        marginBottom: '10px',
        boxSizing: 'border-box',
        backgroundColor: 'lightgrey'
    };

    const buttonStyles = {
        width: 'calc(50% - 0px)',
        marginTop: '15px',
        boxSizing: 'border-box',
        color: 'white',
        cursor: 'pointer'
    };

    const titleStyles = {
        border: '5px solid #282c34',
        width: '100%',
        height: '30px',
        color: 'white',
        margin: '10px 0 20px 0',
        textAlign: 'center',
        backgroundColor: '#007bff'
    };

    const errorStyles = {
        width: '100%',
        textAlign: 'center',
        color: 'red',
        fontSize: '15px',
    };

    const [name, setName] = useState('');
    const [description, setDescription] = useState('');
    const [color, setColor] = useState('#000000');

    const cancelForm = () => {
        displayMarkers();
        getKins();
        setShowNewForm(false);
    };

    const handleChange = (event) => {
        const { name, value } = event.target;
        if (name === 'name') {
            setName(value);
        } else if (name === 'description') {
            setDescription(value);
        } else if (name === 'color') {
            setColor(value);
        }
    };

    const handleSubmit = async (event) => {
        event.preventDefault();

        const formData = {
            kinWorld: worldId,
            kinName: name,
            kinDescription: description,
            kinColor: color,
        };
        console.log(formData);
        try {
            const response = await axios.post('https://loremapper-backend-b042c39916b5.herokuapp.com/kin/create', formData);
            const kinData = response.data.kinData;
            if (!kinData.created) {
                document.getElementById('formError').innerHTML = kinData.error;
            } else {
                cancelForm();
            }
        } catch (error) {
            console.error('Create error: ', error);
            document.getElementById('formError').innerHTML = error.message;
        }
    };

    useEffect(() => {
        return () => {
            cancelForm();
        };
    }, []);

    return (
        <div style={divStyles}>
            <h3 style={titleStyles}>Nueva cultura</h3>
            <form onSubmit={handleSubmit} style={formStyles} id='kinForm'>
                <label style={labelStyles}>Nombre:</label>
                <input style={inputStyles} id='kinNameForm' type="text" name="name" value={name} maxLength={100} onChange={handleChange} />
                <label style={labelStyles}>Descripci√≥n:</label>
                <textarea style={{ ...inputStyles, resize: 'vertical' }} id='kinDescriptionForm' name="description" value={description} maxLength={1000} onChange={handleChange}></textarea>
                <label style={labelStyles}>Color:</label>
                <input style={inputStyles} id='kinColorForm' type="color" name="color" value={color} onChange={handleChange} />
                <button style={{ ...buttonStyles, backgroundColor: 'red', color: 'white', fontWeight: 'bold' }} type="button" onClick={cancelForm}>Cancelar</button>
                <button style={{ ...buttonStyles, backgroundColor: '#007bff', color: 'white', fontWeight: 'bold' }} id='submitKinButton' type="submit">Confirmar</button>
                <span id="formError" style={errorStyles}></span>
            </form>
        </div>
    );
};

export default NewKin;