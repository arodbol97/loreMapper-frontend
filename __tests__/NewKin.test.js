import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import axios from 'axios';
import NewKin from '../src/components/NewKin';

jest.mock('axios');


describe('NewKin', () => {

    const WORLD_ID = 1;        
    const POST_RESPONSE = {
        "data": {
            "kinData": {
                "created": true
            }
        }
    }

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should call axios with the correct data when form is submitted', async () => {
        const displayMarkers = jest.fn();
        const setShowNewForm = jest.fn();
        const getKins = jest.fn();
        axios.post.mockResolvedValue(POST_RESPONSE);
        const { container} = render(<NewKin worldId={WORLD_ID} displayMarkers={displayMarkers} setShowNewForm={setShowNewForm} getKins={getKins}/>);
        const nameInput = container.querySelector('#kinNameForm');
        const descriptionInput = container.querySelector('#kinDescriptionForm');
        const colorInput = container.querySelector('#kinColorForm');        
        const submitButton = container.querySelector('#submitKinButton');
        const URL = 'https://loremapper-backend-b042c39916b5.herokuapp.com/kin/create'
        const EXPECTED_REQUEST_DATA = {
            kinWorld: WORLD_ID,  
            kinName: 'Kin Name',            
            kinDescription: "Kin Description",
            kinColor: "#000000"
        };
    
        fireEvent.change(nameInput, { target: { value: 'Kin Name' } });
        fireEvent.change(descriptionInput, { target: { value: 'Kin Description' } });
        fireEvent.change(colorInput, { target: { value: '#000000' } });        
        fireEvent.click(submitButton);
    
        await waitFor(() => {
          expect(axios.post).toHaveBeenCalledTimes(1);
          expect(axios.post).toHaveBeenCalledWith(URL, EXPECTED_REQUEST_DATA);
        });
      });



      it('should match snapshot', async () => {
        const displayMarkers = jest.fn();
        const setShowNewForm = jest.fn();
        const getKins = jest.fn();
        const { asFragment } = render(<NewKin worldId={WORLD_ID} displayMarkers={displayMarkers} setShowNewForm={setShowNewForm} getKins={getKins}/>);
        expect(asFragment()).toMatchSnapshot();
      });

});
