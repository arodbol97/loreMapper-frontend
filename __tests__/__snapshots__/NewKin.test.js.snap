// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewKin should match snapshot 1`] = `
<DocumentFragment>
  <div
    style="display: flex; flex-wrap: wrap; width: 100%; align-items: flex-start;"
  >
    <h3
      style="border: 5px solid #282c34; width: 100%; height: 30px; color: white; margin: 10px 0px 20px 0px; text-align: center; background-color: rgb(0, 123, 255);"
    >
      Nueva cultura
    </h3>
    <form
      id="kinForm"
      style="width: 100%;"
    >
      <label
        style="color: white; width: 20%; text-align: right; padding-right: 10px; margin-bottom: 5px;"
      >
        Nombre:
      </label>
      <input
        id="kinNameForm"
        maxlength="100"
        name="name"
        style="width: calc(100% - 0px); margin-bottom: 10px; box-sizing: border-box; background-color: lightgrey;"
        type="text"
        value=""
      />
      <label
        style="color: white; width: 20%; text-align: right; padding-right: 10px; margin-bottom: 5px;"
      >
        Descripci√≥n:
      </label>
      <textarea
        id="kinDescriptionForm"
        maxlength="1000"
        name="description"
        style="width: calc(100% - 0px); margin-bottom: 10px; box-sizing: border-box; background-color: lightgrey; resize: vertical;"
      />
      <label
        style="color: white; width: 20%; text-align: right; padding-right: 10px; margin-bottom: 5px;"
      >
        Color:
      </label>
      <input
        id="kinColorForm"
        name="color"
        style="width: calc(100% - 0px); margin-bottom: 10px; box-sizing: border-box; background-color: lightgrey;"
        type="color"
        value="#000000"
      />
      <button
        style="width: calc(50% - 0px); margin-top: 15px; box-sizing: border-box; color: white; cursor: pointer; background-color: red; font-weight: bold;"
        type="button"
      >
        Cancelar
      </button>
      <button
        id="submitKinButton"
        style="width: calc(50% - 0px); margin-top: 15px; box-sizing: border-box; color: white; cursor: pointer; background-color: rgb(0, 123, 255); font-weight: bold;"
        type="submit"
      >
        Confirmar
      </button>
      <span
        id="formError"
        style="width: 100%; text-align: center; color: red; font-size: 15px;"
      />
    </form>
  </div>
</DocumentFragment>
`;
